### DISCUSSION TITLE
State update in the background without screen refresh in React app using Zustand

### DISCUSSION THREAD

**Discussion by lukaszkowalik2** [pmndrs/zustand#1980] [Q&A] [answered]
I am facing an issue with my React application using Zustand (v4.3.7) for state management (React v18.2.0). The problem occurs when the state is updated in the background, but the screen does not refresh to reflect the changes after the first update. Subsequent state updates do not trigger a screen refresh, and I expect the screen to update every time the state changes.

Expected Behavior:
I expect that when the state is updated using Zustand, the changes should be immediately reflected on the screen after each update, causing the screen to refresh accordingly.

Expected Behavior:
I expect that when the state is updated using Zustand (v4.3.7), the changes should be immediately reflected on the screen after each update. The screen should refresh automatically, showing the updated state without any manual intervention or additional user interactions. For example, if I click on a button that triggers a state update, I anticipate that the screen will display the updated state.

Actual Behavior:
Currently, the state updates correctly in the background after the initial update. However, the screen does not automatically refresh to display the subsequent state changes. As a result, after the first update, the screen remains static, and any further state updates do not trigger a screen refresh or reflect the changes.

Steps to Reproduce:

Open the provided [Codesandbox link](https://codesandbox.io/s/still-https-8nt7zf?file=/src/Order.tsx).
Click on the "Plus" button multiple times to trigger state updates, increasing the product quantity.
Notice that the screen refreshes and displays the updated quantity only for the first click. Subsequent clicks on the "Plus" or "Minus" button do not trigger a screen refresh or reflect the changes in the displayed quantity.

Additional Information:

React version: 18.2.0
Zustand version: 4.3.7
Relevant code snippets: If possible, include relevant parts of the code where you are updating the state using Zustand.
Sandbox Link: [Codesandbox link](https://codesandbox.io/s/still-https-8nt7zf?file=/src/Order.tsx) (Please find the code and reproduce the issue in this sandbox environment)
Please note that I am relatively new to Zustand, so any suggestions or solutions with detailed explanations will be highly appreciated. Thank you in advance for your assistance!

Please note that I am relatively new to Zustand, so any suggestions or solutions with detailed explanations will be highly appreciated. Thank you in advance for your assistance!

**Comment by kangju2000** [comment]
This is not a Zustand issue.

It's because `updatedOrders` is doing a shallow copy of state with a spread operator, and it's not copying up to the nested array of objects called products.

To update the state correctly, you need to perform a **deep copy**.

**Comment by lukaszkowalik2** [comment]
Thank you for looking into the issue I reported regarding the state update problem in my React app using Zustand. I appreciate your response and the explanation provided.

I tried implementing the solution you suggested by performing a deep copy of the state instead of a shallow copy using the spread operator. However, even after making the changes in my Codesandbox, the issue persists, and the screen still does not refresh after subsequent state updates.

To provide you with more context, I have updated my Codesandbox with the changes I made based on your suggestion. You can find the updated Codesandbox here: [Codesandbox link](https://codesandbox.io/s/still-https-8nt7zf?file=/src/store.ts).

I was wondering if you could take a look at the updated Codesandbox and see if there's something I might have missed or if there's an alternative approach to correctly performing the deep copy of the state with Zustand.

Additionally, if possible, could you try updating my Codesandbox with the correct solution so that I can see the changes and understand better how to address this issue properly?

Again, thank you for your assistance, and I'm looking forward to your response.

**Reply by kangju2000** [reply, marked as answer]
[Codesandbox](https://codesandbox.io/s/infallible-surf-nt6ppg?file=/src/Order.tsx)

```ts
//before
const fetchOrder = useOrdersStore((state) => state.fetchOrder);
const order = fetchOrder(currentOrderId);

//after
const order = useOrdersStore((state) => state.fetchOrder(currentOrderId));

```

**Reply by lukaszkowalik2** [reply]
Thank you so much for resolving my problem! I truly appreciate your guidance. Thanks again for your help!
